tools/autograd/templates/Functions.cpp

```diff
@@ -488,20 +489,26 @@ Tensor repeat_backward(Tensor grad, int64_t input_dims, IntList repeats) {
   return grad;
 }
 
-Tensor select_backward_scalar(Tensor grad, const Tensor & input, const Tensor & value) {
+Tensor select_equals_backward(Tensor grad, const Tensor & input, const Tensor & value) {
   auto grad_input = zeros_like(input);
   grad_input.masked_fill_(input == value, grad);
   return grad_input;
 }
 
-Tensor select_backward(Tensor grad, int64_t dim, Tensor indices, IntList sizes, bool keepdim) {
+Tensor index_select_backward(Tensor grad, int64_t dim, Tensor indices, IntList sizes, bool keepdim) {
   if (!keepdim) {
     grad = grad.unsqueeze(dim);
     indices = indices.unsqueeze(dim);
   }
   return at::zeros(sizes, grad.type()).scatter_(dim, indices, grad);
 }
 
+Tensor slice_backward(Tensor grad, IntList input_sizes, int64_t dim, int64_t start, int64_t end, int64_t step) {
+  auto grad_input = at::zeros(input_sizes, grad.type());
```

Comment:
Use `grad.options()` in the future, it preserves the device

Reply:
thanks!

Reply:
but i thought this will be guarded by `DeviceGuard` anyways?

Reply:
`.options()` does seem better. I'm just curious.

Reply:
When you pass `.type()` you construct `TensorOptions(Type)` which sets the device index to -1 if `Type` is a CUDA type, so `grad_input` here is always allocated on the current device, whatever that is, while `grad.options()` would means `grad_input` is allocated on the device of `grad`